lib/Doctrine/ORM/UnitOfWork.php

```diff
@@ -2401,6 +2401,13 @@ public function clear($entityName = null)
                     $this->doDetach($entity, $visited, false);
                 }
             }
+
+            foreach ($this->entityInsertions as $hash => $entity) {
```

Comment:
This block of code (starting at line [2392](https://github.com/doctrine/doctrine2/pull/5850/files#diff-6e8c1c1e78b054ba05e20ea09d877865R2392)) seems to become too complex. Can you move it all into its own private method?